llvm/lib/Target/AIE/AIEBaseISelLowering.cpp

```diff
@@ -351,6 +365,23 @@ bool AIEBaseTargetLowering::isEligibleForTailCallOptimization(
   return true;
 }
 
+LLT AIEBaseTargetLowering::getOptimalMemOpLLT(
+    const MemOp &Op, const AttributeList &FuncAttributes) const {
+
+  if (Subtarget.isAIE2()) {
+    if (AlloVecRegMemOps && Op.size() >= 32 && Op.isAligned(Align(32)))
+      return LLT::fixed_vector(8, 32);
+    if (AlloVecRegMemOps && Op.size() >= 16 && Op.isAligned(Align(16)))
+      return LLT::fixed_vector(4, 32);
```

Comment:
There is an interesting explanation. If we look the memset test, we will see (not the same type, but same idea):

```
; CHECK-NEXT:    nopb ; mova r0, #0; nops ; nopxm ; nopv
; CHECK-NEXT:    nopa ; vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
; CHECK-NEXT:    vpush.lo.32 x0, r0, x0
```

If we choose (16, 16) we need 2x more vpush operations (CombinerHelper).

Reply:
`LLT::fixed_vector(4, 32)` represents `<4 x s32>`, i.e. 32-bit elements, not 4-bit.

The memset code is still quite inefficient. The `VPUSH` instructions are produced from the custom legalization of G_BUILD_VECTOR in our legalizer. For vector splats, we actually have an instruction `VBCST` that can be used instead.
That could be added as a standalone pre-legalizer combiner in a follow-up PR.

In your test `lowerMemsetUsingWordVector32` you store 384-bits. That is broken down into a 1 x 256-bit vector store and 4 x 32-bit scalar stores.
Ideally the remaining 4 x 32-bit scalar would be stored as a 128-bit vector. That could then reuse the LSBs of the 256 vector.

I think the optimal code could look as follows:

```
lowerMemsetUsingWordVector32:
mova r0, #0 ; ret lr
movxm p0, #buffer1   // Delay Slot 5
vbcst.32 wl0, r0     // Delay Slot 4
vst wl0, [p0], #32;  // Delay Slot 3
vst.128 wl0, [p0]    // Delay Slot 2
nop                  // Delay Slot 1
```

Not sure how difficult it would be to generate this though :)

Reply:
Yeah, I think broadcast should be the base for memset.

Reply:
I think this could be improved in two different follow-up works:

1. For the first case, we could have a pre-legalization combiner that tracks chains of dependent `VPUSHs` and try to convert them to `VBCST`.
2. For the second case it is more tricky, part of solution is to improve `findGISelOptimalMemOpLowering` to query que target also for the remaining smaller part of the operation. The implementation calls `getOptimalMemOpLLT` just once.

Right now, most of uses of this code are related to scalar alignments, so I expect not much gain, nerveless, we should investigate.

Reply:
It looks like that:

```
bb.1.entry:
  %0:_(s32) = G_CONSTANT i32 0
  %1:_(p0) = G_GLOBAL_VALUE @str1
  %2:_(s8) = G_CONSTANT i8 0
  %3:_(s32) = G_CONSTANT i32 16
  %4:_(s20) = G_TRUNC %3:_(s32)
  G_MEMSET %1:_(p0), %2:_(s8), %4:_(s20), 1 :: (store (s8) into @str1, align 16)
  $r0 = COPY %0:_(s32)
  PseudoRET implicit $lr, implicit $r0

# End machine code for function main.

# *** IR Dump After AIE2PreLegalizerCombiner (aie2-prelegalizer-combiner) ***:
# Machine code for function main: IsSSA, TracksLiveness

bb.1.entry:
  %0:_(s32) = G_CONSTANT i32 0
  %1:_(p0) = G_GLOBAL_VALUE @str1
  %5:_(s32) = G_CONSTANT i32 0
  %6:_(<4 x s32>) = G_BUILD_VECTOR %5:_(s32), %5:_(s32), %5:_(s32), %5:_(s32)
  G_STORE %6:_(<4 x s32>), %1:_(p0) :: (store (<4 x s32>) into @str1)
  $r0 = COPY %0:_(s32)
  PseudoRET implicit $lr, implicit $r0
```

`G_BUILD_VECTOR` can have a better handling for constants.

Reply:
Yes we should introduce a pre-legalizer combiner that turns the `G_BUILD_VECTOR` with identical elements (aka vector splat) into a new `G_AIE_VBCST` instruction. Doesn't have to be limited to constants only, it works for any  `G_BUILD_VECTOR` with identical elements

Reply:
I think we can do that in a similar fashion to the combiner for `G_AIE_PAD_VECTOR_UNDEF`. Let's create a follow-up story, I think this is beyond the scope of this PR.

Reply:
Created.